### ISSUE TITLE
Compilation error when trying to return a local `ref` variable with ref function intent

### ISSUE THREAD

**Issue by twesterhout** [chapel-lang/chapel#19223]
### Steps to Reproduce

**Source Code:**

```chapel
class Foo {
  var A : [0 ..# 5] int;

  proc getChunk(n : int) ref {
    ref r = A[0 ..# n];
    return r;
  }
}

var foo = new Foo();
writeln(foo.getChunk(4));
```

**Compile command:**

`chpl error.chpl`

This fails with the following error
```
error: Reference to scoped variable r cannot be returned
```

### Configuration Information

- Output of `chpl --version`: 1.26.0 pre-release (9382e47)

Related to #19221

**Comment by bradcray** [comment]
At first, I was worried that this bug indicated that we generally were bad at returning local `ref` variables by `ref` intent in general, but that seems not to be the case as indicated by this test, which works as I'd expect:

```chapel
var x = 42;
proc foo() ref {
  ref y = x;
  return y;
}

foo() = 44;
writeln(x);  // prints 44
```

Speaking of which, here's an even simpler version of the OP:

```chapel
var A: [1..10] real;

proc foo() ref {
  ref X = A[2..3];
  return X;
}

foo() = 2.0;

writeln(A);
```

So I'm guessing that what's happening in the issue in the OP is that we essentially implement array slices as a new local array variable that shares data with the original array, along the lines of:

```chapel
var arraySliceTemp = createSlice(A, 0 ..# n);
ref r = arraySliceTemp;
```

and so the complaint is more about the fact that we're returning a reference to `arraySliceTemp` which is about to go out of scope than due to the reference `r` itself (?).  And to that end, the complaint is accurate in the sense that `arraySliceTemp` will go out of scope at the end of the procedure and cause problems if referenced thereafter (as indicated by compiling this with `--ignore-errors` and seeing things crash).

So I think the fix will need to involve either preserving the local array view temporary beyond the procedure's scope or recognizing this "ref to array view" case and changing it to an `out` intent to copy the descriptor out.  In weighing those approaches, we should refresh our memories as to how `ref ArrView = A[2..3];` is implemented in general, where I'm guessing that there's special-case code for this array slice case as compared to the `ref y = x;` case of integers in my first example.

**Comment by twesterhout** [comment]
> So I'm guessing that what's happening in the issue in the OP is that we essentially implement array slices as a new local array variable that shares data with the original array, along the lines of:

And what's worth, a statement such as `ref ArrView = A[2..3]` actually causes a heap allocation which seems very counter-intuitive to me if it's called a reference (I've just spent some time tracking and removing such allocations, because they were affecting performance...)

**Comment by bradcray** [comment]
> a statement such as ref ArrView = A[2..3] actually causes a heap allocation

This is true.  All of our array implementations, including array views like this one, are implemented using classes which currently require a heap allocation to create the object.

**Comment by twesterhout** [comment]
I just hit this issue again working on a difference piece of code... Are there any update on it? Or perhaps suggestions for a workaround? As @bradcray explained, `arraySliceTemp` is created automatically by the compiler. Is there a chance that I could lay my hands on this internal `createSlice` function and then just return whatever it returns?

**Comment by bradcray** [comment]
Hi Tom — Sorry, this never got attention beyond the original diagnosis, though I'm not sure why.  I must've thought you had a workaround already, or else it just slipped my mind.

I haven't come up with any workarounds that are particularly satisfying, either:
a) doing a deep copy of the array chunk (which I'm guessing you don't want to pay for)
b) returning the things required for the callsite to do the slice itself, as in:

```chapel
class Foo {
  var A : [0 ..# 5] int = [i in 0..<5] i;

  proc getChunk(n : int, out lo, out hi) ref {
    lo = 0;
    hi = 4;
    return A;
  }
}

var foo = new Foo();
var lo, hi;
ref Arr = foo.getChunk(4, lo, hi);
foo.A = -foo.A;
writeln(Arr[lo..hi]);
```

I'll use your wake-up call as an excuse to look into what fixing this in the compiler would take.

**Comment by twesterhout** [comment]
Thank you for the quick reply, Brad! I've also tried to answer my own question, and the following seems to work:
```chapel
use ArrayViewSlice;

class C {
  var arr : [0 ..# 5] int = [i in 0 ..# 5] i;

  pragma "reference to const when const this"
  pragma "fn returns aliasing array"
  proc getChunk(lo : int, hi : int) {
    pragma "no auto destroy" var d = {lo .. hi};
    d._value._free_when_no_arrs = true;
    d._value.definedConst = true;
    var a = new unmanaged ArrayViewSliceArr(
        eltType=arr.eltType,
        _DomPid=d._pid, dom=d._instance,
        _ArrPid=arr._pid, _ArrInstance=arr._value);
    d._value.add_arr(a, locking=false, addToList=false);
    return _newArray(a);
  }
}

proc main() {
  var c = new C();
  writeln(c.arr);
  ref arr = c.getChunk(0, 2);
  arr[1] = -1;
  writeln(arr);
  writeln(c.arr);
}
```
Although I'm guessing as a user I'm not supposed to do that :laughing:

**Comment by bradcray** [comment]
Oh wow, that's very fancy of you!  (and sketches out the sort of thing our compiler's likely to have to do)  Have you checked to make sure it's not leaking memory and valgrind-/asan-safe?

**Comment by twesterhout** [comment]
This is actually just a copy-paste from one of Chapel's internal modules: https://github.com/chapel-lang/chapel/blob/86ac3dfb634c79bd8b62873824b4ed0cb2f03916/modules/internal/ChapelArray.chpl#L1239

Good point! Valgrind does complain about some leak:
```
==1865495== LEAK SUMMARY:
==1865495==    definitely lost: 48 bytes in 1 blocks
==1865495==    indirectly lost: 0 bytes in 0 blocks
==1865495==      possibly lost: 0 bytes in 0 blocks
==1865495==    still reachable: 1,584 bytes in 60 blocks
==1865495==         suppressed: 0 bytes in 0 blocks
```
but these numbers seem independent of the array size (I've tried increasing it from 5 to 50) and number of calls to `getChunk` (I've tried running it in a loop 10 times). Furthermore, using `c.arr[0 .. 2]` instead of `c.getChunk(0, 2)` results in the same leak, so I'm guessing it's something internal rather than my code.

**Comment by twesterhout** [comment]
It's been a while, and I still have the unsafe `new unmanaged ArrayViewSliceArr` calls in my code, so I'm wondering whether there's now a better way to return a slice of an array without copying.
